### PR TITLE
Split field manager with interface

### PR DESCRIPTION
**Pull request by jennybuckley**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Make the field manager type more unit testable

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/cc @apelisse 

/priority backlog
/wg apply

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -826,6 +826,7 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 			if err != nil {
 				return nil, err
 			}
+			reqScope.FieldManager = fieldmanager.NewLazyFieldManager(reqScope.FieldManager, reqScope.Creater, reqScope.Kind)
```
Comment (apelisse) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: I like that we have a different set of dependencies for the lazy one than for the non-lazy one. That's good. Avoids having an increasing number of members in the regular one.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ type fakeObjectDefaulter struct{}
 
 func (d *fakeObjectDefaulter) Default(in runtime.Object) {}
 
-func NewTestFieldManager() *fieldmanager.FieldManager {
+func NewTestFieldManager() fieldmanager.FieldManager {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: That's conflicting with what I'm doing :cry:

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lazy_fieldmanager.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	liveObjAccessor, err := meta.Accessor(liveObj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
+	}
+	if len(liveObjAccessor.GetManagedFields()) == 0 {
+		emptyObj, err := f.objectCreater.New(f.gvk)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create empty object of type %v: %v", f.gvk, err)
+		}
+		liveObj, err = f.fieldManager.Update(emptyObj, liveObj, "before-first-apply")
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lazy_fieldmanager.go]: That's so much better

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lazy_fieldmanager_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+type fakeObjectCreater struct{}
+
+var _ runtime.ObjectCreater = &fakeObjectCreater{}
+
+func (_ *fakeObjectCreater) New(_ schema.GroupVersionKind) (runtime.Object, error) {
+	return &unstructured.Unstructured{Object: map[string]interface{}{}}, nil
+}
+
+func TestUpdateOnlyDoesNotTrackManagedFields(t *testing.T) {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lazy_fieldmanager_test.go]: What about testing that apply starts tracking?
Conflicts with past updated fields?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go (hunk 2 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: I can rebase on top of yours

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lazy_fieldmanager.go [hunk 5]
```diff
@@ -44,28 +43,15 @@ func NewLazyFieldManager(fieldManager FieldManager, objectCreater runtime.Object
...
-			return nil, fmt.Errorf("failed to decode managed fields: %v", err)
-		}
+	newObjAccessor, err := meta.Accessor(newObj)
+	if err != nil {
+		return newObj, nil
 	}
-
-	// if managed field is still empty, skip updating managed fields altogether
-	if len(managed.Fields) == 0 {
+	if len(liveObjAccessor.GetManagedFields()) == 0 && len(newObjAccessor.GetManagedFields()) == 0 {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lazy_fieldmanager.go]: I think that's cool

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/BUILD [hunk 6]
```diff
@@ -2,7 +2,10 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "go_default_library",
-    srcs = ["fieldmanager.go"],
+    srcs = [
+        "fieldmanager.go",
+        "lazy_fieldmanager.go",
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/BUILD]: I don't know if I'm a big fan of the name. "Lazy" doesn't really say why it's being lazy. What about: `TrackAppliedFieldManager`, that's pretty bad, but you get the idea.

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 7]
```diff
@@ -560,7 +560,7 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		if err != nil {
 			return nil, fmt.Errorf("failed to create field manager: %v", err)
 		}
-		reqScope.FieldManager = fm
+		reqScope.FieldManager = fieldmanager.NewLazyFieldManager(fm, a.group.Creater, fqKindToRegister)
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: Why don't we create the lazyfieldmanager at the same time that we create the fieldmanager?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (apelisse) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: I find it a little confusing that you put the fieldmanager into reqScope and then replace it right away, maybe use a temporary instead?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/BUILD (hunk 6 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/BUILD]: I'll think about it

Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/BUILD]: Maybe SkipNonAppliedFieldManager?

**Comment by apelisse** [comment]
Thank you Jenny!
/lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go [hunk 8]
```diff
@@ -172,15 +179,13 @@ func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (r
...
 }
 
-// Apply is used when server-side apply is called, as it merges the
-// object and update the managed fields.
-func (f *FieldManager) Apply(liveObj runtime.Object, patch []byte, fieldManager string, force bool) (runtime.Object, error) {
+// Apply implements FieldManager.
+func (f *fieldManager) Apply(liveObj runtime.Object, patch []byte, fieldManager string, force bool) (runtime.Object, error) {
 	// If the object doesn't have metadata, apply isn't allowed.
-	accessor, err := meta.Accessor(liveObj)
+	_, err := meta.Accessor(liveObj)
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go]: Is it still useful to get / check this error?

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go]: I guess Update does it too; nit: might as well do it all in one line as in that function.

**Comment by lavalamp** [comment]
/approve

nit is mostly optional.

**Pull request merged**